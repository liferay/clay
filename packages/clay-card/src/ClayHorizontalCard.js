import 'clay-checkbox';
import 'clay-link';
import 'clay-sticker';
// eslint-disable-next-line
import { ClayActionsDropdown } from 'clay-dropdown';
import {Config} from 'metal-state';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';

import actionItemsValidator from './action_items_validator';
import templates from './ClayHorizontalCard.soy.js';

/**
 * Metal ClayHorizontalCard component.
 */
class ClayHorizontalCard extends Component {
	/**
	 * Continues the propagation of the checkbox changed event
	 * @param {!Event} event
	 * @private
	 */
	_handleItemCheckboxClick(event) {
		this.emit('itemToggled', event);
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayHorizontalCard.STATE = {
	/**
	 * List of items to display in the actions menu.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?array|undefined}
	 * @default undefined
	 */
	actionItems: actionItemsValidator,

	/**
	 * Flag to indicate if the card is disabled or not.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?bool}
	 * @default false
	 */
	disabled: Config.bool().value(false),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Group Index the card belongs to.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	groupName: Config.string(),

	/**
	 * Url to be applied to the title of the card.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	href: Config.string(),

	/**
	 * Icon of the card.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string}
	 * @default folder
	 */
	icon: Config.string().value('folder'),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * Name to be applied to the input element.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	inputName: Config.string(),

	/**
	 * Value to be applied to the input element.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	inputValue: Config.string(),

	/**
	 * Flag to indicate if the card is selectable or not.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?bool}
	 * @default false
	 */
	selectable: Config.bool().value(false),

	/**
	 * Flag to indicate if the card is selected or not.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?bool}
	 * @default false
	 */
	selected: Config.bool().value(false),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),

	/**
	 * Title of the card.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {!string}
	 * @default undefined
	 */
	title: Config.string().required(),
};

defineWebComponent('clay-horizontal-card', ClayHorizontalCard);

Soy.register(ClayHorizontalCard, templates);

export {ClayHorizontalCard};
export default ClayHorizontalCard;
