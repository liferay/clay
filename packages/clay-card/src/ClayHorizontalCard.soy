{namespace ClayHorizontalCard}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param spritemap: string}
	{@param title: string}
	{@param? actionItems: list<[
		href: string,
		label: string,
		separator: bool
	]>}
	{@param? disabled: bool}
	{@param? elementClasses: string}
	{@param? groupName: string}
	{@param? handleItemCheckboxClick_: any}
	{@param? href: string}
	{@param? icon: string}
	{@param? id: string}
	{@param? inputName: string}
	{@param? inputValue: string}
	{@param? selectable: bool}
	{@param? selected: bool}

	{let $attributes kind="attributes"}
		class="card-type-directory
			{if $selectable}
				{sp}form-check form-check-card form-check-middle-left
			{else}
				{sp}card card-horizontal
			{/if}

			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	{let $descriptiveArea kind="html"}
		{call .descriptiveArea}
			{param actionItems: $actionItems /}
			{param href: $href /}
			{param icon: $icon /}
			{param spritemap: $spritemap /}
			{param title: $title /}
		{/call}
	{/let}

	<div {$attributes}>
		{if $selectable}
			{call ClayCheckbox.render}
				{param ariaLabelledBy: $groupName ? 'group-' + $groupName : null /}
				{param checked: $selected /}
				{param disabled: $disabled /}
				{param events: ['change': $handleItemCheckboxClick_] /}
				{param labelContent kind="html"}
					<div class="card card-horizontal">
						{$descriptiveArea}
					</div>
				{/param}

				{param name: $inputName /}
				{param value: $inputValue /}
			{/call}
		{else}
			{$descriptiveArea}
		{/if}
	</div>
{/template}

/**
 * The content the element.
 */
{template .descriptiveArea}
	{@param spritemap: string}
	{@param title: string}
	{@param? actionItems: list<[
		href: string,
		label: string,
		separator: bool
	]>}
	{@param? href: string}
	{@param? icon: string}

	<div class="card-body">
		<div class="card-row">
			{let $classesDescription kind="text"}
				autofit-col autofit-col-expand autofit-col-gutters
			{/let}

			<div class="autofit-col">
				{call ClaySticker.render}
					{param icon: $icon ?: 'folder' /}
					{param spritemap: $spritemap /}
					{param style: 'unstyled' /}
				{/call}
			</div>

			<div class="{$classesDescription}">
				{if $href}
					{call ClayLink.render}
						{param elementClasses: 'card-title text-truncate' /}
						{param href: $href /}
						{param label: $title /}
					{/call}
				{else}
					<div class="card-title text-truncate">{$title}</div>
				{/if}
			</div>

			{if $actionItems}
				<div class="autofit-col">
					{call ClayActionsDropdown.render}
						{param items: $actionItems /}
						{param spritemap: $spritemap /}
					{/call}
				</div>
			{/if}
		</div>
	</div>
{/template}
