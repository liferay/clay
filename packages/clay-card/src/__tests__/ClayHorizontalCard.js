import ClayHorizontalCard from '../ClayHorizontalCard';

let component;
let consoleErrorReference;
let spritemap = 'icons.svg';
let actionItems = [
	{
		href: '#1',
		label: 'Edit',
	},
	{
		href: '#1',
		label: 'Save',
	},
];

/**
 * Stubs console.error
 */
function mockConsoleError() {
	console.error = () => {};
}

describe('ClayHorizontalCard', function() {
	beforeEach(() => {
		consoleErrorReference = console.error;
	});

	afterEach(() => {
		if (component) {
			component.dispose();
		}

		console.error = consoleErrorReference;
	});

	it('should render the default markup', () => {
		component = new ClayHorizontalCard({
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with classes', () => {
		component = new ClayHorizontalCard({
			elementClasses: 'my-custom-class',
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with id', () => {
		component = new ClayHorizontalCard({
			id: 'myId',
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with actionItems', () => {
		component = new ClayHorizontalCard({
			actionItems: actionItems,
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selectable ClayHorizontalCard', () => {
		component = new ClayHorizontalCard({
			selectable: true,
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selected ClayHorizontalCard', () => {
		component = new ClayHorizontalCard({
			selectable: true,
			selected: true,
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a disabled ClayHorizontalCard', () => {
		component = new ClayHorizontalCard({
			disabled: true,
			selectable: true,
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selectable ClayHorizontalCard with input `name`', () => {
		component = new ClayHorizontalCard({
			inputName: 'checkbox01',
			selectable: true,
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selectable ClayHorizontalCard with input `value`', () => {
		component = new ClayHorizontalCard({
			inputValue: 'checkbox',
			selectable: true,
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with href', () => {
		component = new ClayHorizontalCard({
			href: '#1',
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with a different icon', () => {
		component = new ClayHorizontalCard({
			icon: 'list',
			spritemap: spritemap,
			title: 'My Title',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selectable ClayHorizontalCard and emit an event on item toggle', () => {
		component = new ClayHorizontalCard({
			selectable: true,
			spritemap: spritemap,
			title: 'My Title',
		});

		const spy = jest.spyOn(component, 'emit');

		component.element.querySelector('input[type="checkbox"]').click();

		expect(spy).toHaveBeenCalled();
		expect(spy).toHaveBeenCalledWith('itemToggled', expect.any(Object));
	});

	it('should fail when no spritemap is passed', function() {
		mockConsoleError();

		expect(() => {
			component = new ClayHorizontalCard({
				title: 'My Title',
			});
		}).toThrow();
	});

	it('should fail when no title is passed', function() {
		mockConsoleError();

		expect(() => {
			component = new ClayHorizontalCard({
				spritemap: spritemap,
			});
		}).toThrow();
	});
});
