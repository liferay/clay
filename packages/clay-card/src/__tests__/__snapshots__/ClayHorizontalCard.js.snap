// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayHorizontalCard should render a ClayHorizontalCard with a different icon 1`] = `
<div class="card-type-directory card card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <div class="autofit-col">
        <span class="sticker sticker-rounded sticker-unstyled">
          <span class="inline-item">
            <svg aria-hidden="true" class="lexicon-icon lexicon-icon-list" focusable="false">
              <title>list</title>
              <use xlink:href="icons.svg#list"></use>
            </svg>
          </span>
        </span>
      </div>
      <div class="autofit-col autofit-col-expand autofit-col-gutters">
        <div class="card-title text-truncate">My Title</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with actionItems 1`] = `
<div class="card-type-directory card card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <div class="autofit-col">
        <span class="sticker sticker-rounded sticker-unstyled">
          <span class="inline-item">
            <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder" focusable="false">
              <title>folder</title>
              <use xlink:href="icons.svg#folder"></use>
            </svg>
          </span>
        </span>
      </div>
      <div class="autofit-col autofit-col-expand autofit-col-gutters">
        <div class="card-title text-truncate">My Title</div>
      </div>
      <div class="autofit-col">
        <div class="dropdown dropdown-action">
          <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn component-action btn-unstyled" ref="triggerButton" title="actions" type="button">
            <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v" focusable="false">
              <title>ellipsis-v</title>
              <use xlink:href="icons.svg#ellipsis-v"></use>
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="#1">Edit</a>
    </li>
    <li>
      <a class=" dropdown-item" href="#1">Save</a>
    </li>
  </ul>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with classes 1`] = `
<div class="card-type-directory card card-horizontal my-custom-class">
  <div class="card-body">
    <div class="card-row">
      <div class="autofit-col">
        <span class="sticker sticker-rounded sticker-unstyled">
          <span class="inline-item">
            <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder" focusable="false">
              <title>folder</title>
              <use xlink:href="icons.svg#folder"></use>
            </svg>
          </span>
        </span>
      </div>
      <div class="autofit-col autofit-col-expand autofit-col-gutters">
        <div class="card-title text-truncate">My Title</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with href 1`] = `
<div class="card-type-directory card card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <div class="autofit-col">
        <span class="sticker sticker-rounded sticker-unstyled">
          <span class="inline-item">
            <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder" focusable="false">
              <title>folder</title>
              <use xlink:href="icons.svg#folder"></use>
            </svg>
          </span>
        </span>
      </div>
      <div class="autofit-col autofit-col-expand autofit-col-gutters">
        <a class=" card-title text-truncate" href="#1">My Title</a>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with id 1`] = `
<div class="card-type-directory card card-horizontal" id="myId">
  <div class="card-body">
    <div class="card-row">
      <div class="autofit-col">
        <span class="sticker sticker-rounded sticker-unstyled">
          <span class="inline-item">
            <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder" focusable="false">
              <title>folder</title>
              <use xlink:href="icons.svg#folder"></use>
            </svg>
          </span>
        </span>
      </div>
      <div class="autofit-col autofit-col-expand autofit-col-gutters">
        <div class="card-title text-truncate">My Title</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a disabled ClayHorizontalCard 1`] = `
<div class="card-type-directory form-check form-check-card form-check-middle-left">
  <div class="custom-control custom-checkbox">
    <label>
      <input disabled="disabled" class="custom-control-input" ref="input" type="checkbox">
      <span class="custom-control-label"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <div class="autofit-col">
              <span class="sticker sticker-rounded sticker-unstyled">
                <span class="inline-item">
                  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder" focusable="false">
                    <title>folder</title>
                    <use xlink:href="icons.svg#folder"></use>
                  </svg>
                </span>
              </span>
            </div>
            <div class="autofit-col autofit-col-expand autofit-col-gutters">
              <div class="card-title text-truncate">My Title</div>
            </div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a selectable ClayHorizontalCard 1`] = `
<div class="card-type-directory form-check form-check-card form-check-middle-left">
  <div class="custom-control custom-checkbox">
    <label>
      <input class="custom-control-input" ref="input" type="checkbox">
      <span class="custom-control-label"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <div class="autofit-col">
              <span class="sticker sticker-rounded sticker-unstyled">
                <span class="inline-item">
                  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder" focusable="false">
                    <title>folder</title>
                    <use xlink:href="icons.svg#folder"></use>
                  </svg>
                </span>
              </span>
            </div>
            <div class="autofit-col autofit-col-expand autofit-col-gutters">
              <div class="card-title text-truncate">My Title</div>
            </div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a selectable ClayHorizontalCard with input \`name\` 1`] = `
<div class="card-type-directory form-check form-check-card form-check-middle-left">
  <div class="custom-control custom-checkbox">
    <label>
      <input class="custom-control-input" name="checkbox01" ref="input" type="checkbox">
      <span class="custom-control-label"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <div class="autofit-col">
              <span class="sticker sticker-rounded sticker-unstyled">
                <span class="inline-item">
                  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder" focusable="false">
                    <title>folder</title>
                    <use xlink:href="icons.svg#folder"></use>
                  </svg>
                </span>
              </span>
            </div>
            <div class="autofit-col autofit-col-expand autofit-col-gutters">
              <div class="card-title text-truncate">My Title</div>
            </div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a selectable ClayHorizontalCard with input \`value\` 1`] = `
<div class="card-type-directory form-check form-check-card form-check-middle-left">
  <div class="custom-control custom-checkbox">
    <label>
      <input class="custom-control-input" value="checkbox" ref="input" type="checkbox">
      <span class="custom-control-label"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <div class="autofit-col">
              <span class="sticker sticker-rounded sticker-unstyled">
                <span class="inline-item">
                  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder" focusable="false">
                    <title>folder</title>
                    <use xlink:href="icons.svg#folder"></use>
                  </svg>
                </span>
              </span>
            </div>
            <div class="autofit-col autofit-col-expand autofit-col-gutters">
              <div class="card-title text-truncate">My Title</div>
            </div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a selected ClayHorizontalCard 1`] = `
<div class="card-type-directory form-check form-check-card form-check-middle-left">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" class="custom-control-input" ref="input" type="checkbox">
      <span class="custom-control-label"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <div class="autofit-col">
              <span class="sticker sticker-rounded sticker-unstyled">
                <span class="inline-item">
                  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder" focusable="false">
                    <title>folder</title>
                    <use xlink:href="icons.svg#folder"></use>
                  </svg>
                </span>
              </span>
            </div>
            <div class="autofit-col autofit-col-expand autofit-col-gutters">
              <div class="card-title text-truncate">My Title</div>
            </div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render the default markup 1`] = `
<div class="card-type-directory card card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <div class="autofit-col">
        <span class="sticker sticker-rounded sticker-unstyled">
          <span class="inline-item">
            <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder" focusable="false">
              <title>folder</title>
              <use xlink:href="icons.svg#folder"></use>
            </svg>
          </span>
        </span>
      </div>
      <div class="autofit-col autofit-col-expand autofit-col-gutters">
        <div class="card-title text-truncate">My Title</div>
      </div>
    </div>
  </div>
</div>
`;
